File: python-service/app/models/teacher_youtube_token.py
# ============================================================
# FILE 7: app/models/teacher_youtube_token.py
# ============================================================
"""
TeacherYouTubeToken Model - OAuth tokens for YouTube uploads
"""
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from app.core.database import Base
from datetime import datetime


class TeacherYouTubeToken(Base):
    __tablename__ = "teacher_youtube_tokens"
    __table_args__ = {'schema': 'academic'}

    id = Column(Integer, primary_key=True, index=True)

    # ==================== Relationship ====================
    teacher_id = Column(Integer, ForeignKey('academic.users.id'), unique=True, nullable=False)
    
    teacher = relationship("User", back_populates="youtube_token")

    # ==================== OAuth Tokens ====================
    access_token = Column(Text, nullable=False)
    refresh_token = Column(Text, nullable=False)
    token_type = Column(String(20), default='Bearer')

    # ==================== Token Metadata ====================
    expires_at = Column(DateTime(timezone=True), nullable=False)
    scope = Column(Text)

    # ==================== YouTube Channel Info ====================
    youtube_channel_id = Column(String(50))
    youtube_channel_name = Column(String(200))

    # ==================== Audit Fields ====================
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
    last_used_at = Column(DateTime(timezone=True))

    # ==================== Helper Methods ====================
    def is_expired(self):
        return datetime.utcnow() > self.expires_at

    def needs_refresh(self):
        # Refresh if token expires in next 5 minutes
        return datetime.utcnow().timestamp() + 300 > self.expires_at.timestamp()

File: java-service/src/main/java/com/edu/platform/model/TeacherYouTubeToken.java
package com.edu.platform.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;

@Entity
@Table(name = "teacher_youtube_tokens", schema = "academic")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TeacherYouTubeToken {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // ==================== Relationship ====================
    
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "teacher_id", unique = true, nullable = false)
    private User teacher;

    // ==================== OAuth Tokens (should be encrypted in real implementation) ====================
    
    @Column(name = "access_token", nullable = false, columnDefinition = "TEXT")
    private String accessToken;

    @Column(name = "refresh_token", nullable = false, columnDefinition = "TEXT")
    private String refreshToken;

    @Column(name = "token_type", length = 20)
    @Builder.Default
    private String tokenType = "Bearer";

    // ==================== Token Metadata ====================
    
    @Column(name = "expires_at", nullable = false)
    private Instant expiresAt;

    @Column(name = "scope", columnDefinition = "TEXT")
    private String scope;

    // ==================== YouTube Channel Info ====================
    
    @Column(name = "youtube_channel_id", length = 50)
    private String youtubeChannelId;

    @Column(name = "youtube_channel_name", length = 200)
    private String youtubeChannelName;

    // ==================== Audit Fields ====================
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(name = "last_used_at")
    private Instant lastUsedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }

    // ==================== Helper Methods ====================
    
    public boolean isExpired() {
        return Instant.now().isAfter(this.expiresAt);
    }

    public boolean needsRefresh() {
        // Refresh if token expires in next 5 minutes
        return Instant.now().plusSeconds(300).isAfter(this.expiresAt);
    }
}